# Topics 1, 2, 3 and 4

### Q1
```
np.array([[1, 2], [3, 4]])
rank = ???
shape = ???
```

**Rank = 2 and shape = (2, 2)**
```
x = [[1, 2]
     [3, 4]]
```
a $2 \times 2$ matrix. 

---


### Q2
```
np.array([[1, 2], [3, 4]])
x[1] = ???
```

**[3, 4]**

```
a = [1, 2]
b = [3, 4]
x = [a, b]
x[1] = b = [3, 4]
```

---

### Q3
```
input_shape = (60000, 28, 28)
num_samples = ???
```

**60000 samples**

The samples axis is always the first axis. Each sample is a rank 2 tensor of shape ```(28, 28)```.

---


### Q4
```
x = train_images[:128]
num_samples = ???
```


**128 samples**

```a:b``` is a slice from index ```a``` inclusive to ```b``` exclusive.

```[:b]``` is the same as ```[0:b]```

```train[images[0]``` to ```train_images[127]```

---

### Q5
```
second_mini_batch_of_128 = train_images([p:q])
p = ???
q = ???
```


**p = 128, q = 256**

First batch is [0:128], second is [128:256] etc. Batch ```n``` is the slice ```[n * 128:(n + 1) * 128]```. First batch is batch 0, second batch is batch 1 and so on. Remember 0 = first, 1 = second etc.

---


### Q6
```
sigmoid([-1, 0, 1]) = ???
```


**```sigmoid([-1, 0, 1])= [0.5 - a, 0.5, 0.5 + a]``` where $0 < a < 0.5$**

In fact $a = 0.23$.

```sigmoid(0) = 0.5, sigmoid(inf) = 1, sigmoid(-inf) = 0```, rotational symmetry about $(0, 0.5)$

---


### Q7
```
w = [[1, 0],
     [1, 1]],
b = [-2, 2],
x = [-1, 1]
relu(dot(w, x) + b) = ???
```

**```[0, 2]```**

$w \cdot x = [[1, 0] \cdot [-1, 1]^T, [1, 1] \cdot [-1, 1]^T] =  [-1, 0]^T$

$[-1, 0]^T + [-2, 2]^T = [-3, 2]^T$

$\text{relu}([-3, 2]^T) = [0, 2]^T$

---


### Q8
Suppose the third item in a word dictionary is 'deep' and the seventh item is 'learning'. What is the one-hot encoding of 'deep learning'?

**(0, 0, 0, 1, 0, 0, 0, 1, 0, 0...)**

The index 0 is not used. The first dictionary word has index 1, the second word has index 2 and so on. So, place 1's at indices 3 and 7.

---

### Q9

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
model = models.Sequential()
model.add(layers.Dense(64, activation = 'relu', input_shape = (10000,))) # layer 1
model.add(layers.Dense(32, activation = 'relu')) # layer 2

What is the shape of:
1. Layer 1 weight tensor
2. Layer 1 bias tensor
3. Layer 2 weight tensor
4. Layer 2 bias tensor?


In [3]:
print('\033[1m') # print bold
print('1.', model.layers[0].get_weights()[0].shape) # first layer weights
print('2.', model.layers[0].get_weights()[1].shape) # first layer bias
print('3.', model.layers[1].get_weights()[0].shape) # second layer weights
print('4.', model.layers[1].get_weights()[1].shape) # second layer bias
print('\033[0m') # print back to normal


1. (10000, 64)
2. (64,)
3. (64, 32)
4. (32,)



---

### Q10
```
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
model.add(layers.Dense(10, activation = 'softmax'))

```

What problem type has the above network been designed to solve?


**Single label, multi-class**

The final layer has 10 units - there are 10 outputs, so this must be a multi-class problem *and* multi-claaa, sngle layer requires a softmax last layer activation (see DLWP Table 4.1).

---


### Q11
```
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
model.add(layers.Dense(10, activation = 'softmax'))
number_input_elements = ???
```

How many input elements are implied in the above network?

**784 ($\bf 28 \times 28$)**

Inputs to a dense FF network are vectors. The shape of a vector is ```(num_elements, )```.

---

### Q12
```
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28 * 28,))) # layer 1
model.add(layers.Dense(10, activation = 'softmax')) # layer 2

```
1. ```number_of_layer_1_trainable_params = ```
2. ```number_params_layer_2_weight_tensor = ```


1. **```number_of_layer_1_trainable_params = 784 * 512 + 512 = 401920```**

2. **```number_params_layer_2_weight_tensor = 512 * 10 + 10 = 5130```**

In [3]:
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28 * 28,))) # layer 1
model.add(layers.Dense(10, activation = 'softmax')) # layer 2
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_2 (Dense)              (None, 512)               401920    
_________________________________________________________________
dense_3 (Dense)              (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


---

### Q13
```
softmax([1, 10, 1]) = [a, b, c]
```

What can you say about the relative values of $a$, $b$ and $c$?

**$\bf a = c$ and $\bf 0 < a < b$**

Softmax is applied element-wise. All softmax'd elements are greater than zero and the order of sizes is preserved (in fact, amplified).

---

### Q14
```
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
model.add(layers.Dense(x, activation = 'relu'))
model.add(layers.Dense(10, activation = 'softmax'))
```

What is the range of possible value for ```x```, the number of units in the third layer?

$\bf x \ge 10$

Bottlenecks may occur when the number of units of a preceding layer is smaller.

---


### Q15
```
model = models.Sequential()
model.add(layers.Dense(512, activation = 'relu', input_shape = (28*28,)))
model.add(layers.Dense(10, activation = 'softmax'))
model.compile(optimizer = 'rmsprop', loss = unknown_loss, metrics = [unknown_metric])
```

1. ```unknown_loss = ```
2. ```unknown_metric = ```

1. **unknown loss = ```categorical_crossentropy```**

2. **unknown metric = accuracy or ```'acc'```**

This is a single label multi-class problem - refer to DLWP Table 4.1

---

### Q16

Consider ```x``` below.

1. What is the rank of `x`?
2. What is the shape of `x`?
3. What is `x[0]`?
4. What is `x[0][1]`?
5. What is `x[0][1][0]`?

In [5]:
import numpy as np

x = np.array([[[1, 2], [3, 4]], 
              [[1, 0], [0, 1]], 
              [[2, 2], [2, 2]]])

print(x)

[[[1 2]
  [3 4]]

 [[1 0]
  [0 1]]

 [[2 2]
  [2 2]]]


In [7]:
print('\033[1m') # print bold
print('1.', x.ndim)
print('2.', x.shape)
print('3.\n', x[0])
print('4.', x[0][1])
print('5.', x[0][1][0])
print('\033[0m') # print back to normal


1. 3
2. (3, 2, 2)
3.
 [[1 2]
 [3 4]]
4. [3 4]
5. 3



---

### Q17

Consider ```x``` below.

1. What is the rank of `x`?
2. What is the shape of `x`?

In [8]:
x = np.array([1, 2, 3])
print('x = ', x)

x =  [1 2 3]


In [9]:
print('\033[1m') # print bold
print('1.', x.ndim)
print('2.', x.shape)
print('\033[0m') # print back to normal


1. 1
2. (3,)



---

### Q18
Consider the mathematical vector $x = (1, 2, 3)$
1. How many elements does $x$ have?
2. What is the dimensionality of $x$?

1. **3**
2. **3**

---

### Q19

For ```w, b, x``` below,

1. What is $w \cdot x$?
2. What is $w \cdot x + b$?
3. What is ```relu```$(w \cdot x + b)$?

In [11]:
w = np.array([[1, 0, 2], [-2, 1,-1]])
b = np.array([1, 2])
x = np.array([2, 0, 1])

print('w = \n', w)
print('b = ', b)
print('x = ', x)

w = 
 [[ 1  0  2]
 [-2  1 -1]]
b =  [1 2]
x =  [2 0 1]


---

In [12]:
print('\033[1m') # print bold
print('1.', np.dot(w, x))
print('2.', np.dot(w, x) + b)
ans = np.array([5, 0])
print('3.', ans)
print('\033[0m') # print back to normal


1. [ 4 -5]
2. [ 5 -3]
3. [5 0]



### Q20

With regard to the model below:

Fill in the blanks

| Layer (type) | Output Shape | Param # |
|--------------|--------------|---------|
dense_1 (Dense)| (None, ___ ) |  ___    |
dense_2 (Dense)| (None, ___ ) |  ___    |

Total params:____

Trainable params:____


Non-trainable params: ____


In [4]:
from tensorflow.keras import models, layers
model = models.Sequential()
model.add(layers.Dense(512, activation='relu', input_shape=(784, ))) # dense_1
model.add(layers.Dense(10, activation = 'softmax'))                  # dense_2 
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 512)               401920    
                                                                 
 dense_3 (Dense)             (None, 10)                5130      
                                                                 
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


---

### Q21

$x = (1, 2, 1, 1)$. 

1. What is softmax$(x)$?
2. What is the sum of the elements of softmax$(x)$?
3. Is softmax$(x)$ a probability distribution? Why/why not?

In [18]:
import numpy as np
def softmax(x):
    y = np.exp(x)
    return y / np.sum(y)

x = np.array([1, 2, 1, 1])
print('\033[1m') # print bold
print('1.', softmax(x))
print('2.', sum(softmax(x)))
print('3. For all intent and purposes, the output of `softmax` is a probability distribution:')
print('   it preserves the relative weight of the elements of its input, always outputs positive')
print('   numbers, and scales them so that the total always sums to 1.')
print('\033[0m') # print back to normal


1. [0.1748777  0.47536689 0.1748777  0.1748777 ]
2. 1.0000000000000002
3. For all intent and purposes, the output of `softmax` is a probability distribution:
   it preserves the relative weight of the elements of its input, always outputs positive
   numbers, and scales them so that the total always sums to 1.



---

### Q22

An image in the MNIST training is labelled '7'. 

What is the categorical encoding of this label?

$\bf (0, 0, 0, 0, 0, 0, 0, 1, 0, 0)$

---

### Q23

1. What is an affine transformation?
2. What is an activation function?
3. What are the essential properties of an activation function?
4. What is the action of an activated layer on an input vector?
5. Sketch the form of the sigmoid function.


**1. A linear transform followed by a translation i.e. $\bf w \cdot x + b$ where $w$ is the weight matrix $\bf b$ is the bias vector and $\bf x$ is the layer input.**  
**2. An element-wise application of a function to $\bf w.x + b$**  
**3. Non-linearity and differentiability**  
**4. An affine transformation followed by an activation**  
**5. (Fig 3.5)**  


---


### Q24

1. What is a loss function?
2. What is the optimiser?
3. What is gradient descent optimisation?
4. What is stochastic gradient descent?
5. What kind of loss functions are suitable for gradient optimisation?
6. What is backpropagation?


**1. The loss function quantifies the departure of the network prediction from the desired target.**  
**2. The optimiser adjusts weights (and biases) in a direction that will lower the loss for the current sample or mini-batch.**  
**3. Gradient optimisation: weights and biases are moved by a small step in the direction of the negative gradient.**   
**4. In SGD, the loss for a mini-batch, chosen at random, is calculated.**  
**5. Any smooth function such that the loss is 0 when prediction matches target, and otherwise positive.**  
**6. An efficient implementation of the SGD using the chain rule of differentiation.**  


---


# Topic 5

### Q1

```
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
output_shape = (none, x, x, 32)
x = ???
```

**```x``` = 26**

The stride is 1. The $3 \times 3$ filter is moved by one pixel, starting at the first pixel and ending at the 26th (so the filter occupies pixels 26, 27, 28).

In [5]:
from tensorflow.keras import models, layers
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
Total params: 320
Trainable params: 320
Non-trainable params: 0
_________________________________________________________________


---

### Q2

Consider this layer:

```
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
```

1. What is the number of trainable of parameters?  
2. And how many filters?  

1. $\bf (3 * 3 + 1) * 32 = 320$  
2. $\bf 32$  

In [3]:
from tensorflow.keras import models, layers
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_1 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
Total params: 320
Trainable params: 320
Non-trainable params: 0
_________________________________________________________________


---

### Q3

```
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2))
output_shape = (none, x, x, 32)
x = ???
```

**x = 13**

The output shape from the first layer is ```(26, 26)```. The max pooling layer covers the $26 \times 26$ feature map with 2 x 2 patches and each patch produces a single output.

In [9]:
from tensorflow.keras import models, layers
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_3 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
Total params: 320
Trainable params: 320
Non-trainable params: 0
_________________________________________________________________


---

# Topic 6

### Q1
```
model = Sequential()
model.add(Embedding(input_dim=10000, output_dim=3, input_length=50))
```

The output shape of the above ```Embedding``` layer is ```(None, x, y)```.

1. What is x?
2. What is y?
3. How many trainable parameters are there in the ```Embedding``` layer?

1. **x = 50**
2. **y = 3**
3. $\bf input\_dim * output\_dim = 10000 * 3 = 30000$

In [13]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
model = models.Sequential()
model.add(Embedding(input_dim=10000, output_dim=3, input_length=50))
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 50, 3)             30000     
Total params: 30,000
Trainable params: 30,000
Non-trainable params: 0
_________________________________________________________________


---

### Q2 
```
model = Sequential()
model.add(Embedding(input_dim=10000, output_dim=32)))
model.add(SimpleRNN(32))
```

What is the number if trainable RNN_parameters? 

**```32 * 32 + 32 * 32 + 32 = 2080``` or ```(32 + 32 + 1) * 32```**

The Simple RNN layer has two weight tensors, W and U. The layer input vector has 32 elements and this is a 32 unit layer. Therefore, w is $32 \times 32$ and b has 32 elements. The 32 element output feeds back to the layer input so U has $32 \times 32$ elements.

Or: Suppose the layer has M units and the input has N elements. Each unit has $N$ weights for the $N$ input features, $M$ weights for the previous output, and a bias: that's $(N + M + 1)$ coefficients. But there are $M$ units, giving $(N + M + 1) \times M$ coefficients (or, $N \times M + M \times M + M$).


In [15]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding
from tensorflow.keras.layers import SimpleRNN

model = models.Sequential()
model.add(Embedding(input_dim=10000, output_dim=32))
model.add(SimpleRNN(32))
model.summary()

Model: "sequential_9"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, None, 32)          320000    
                                                                 
 simple_rnn_2 (SimpleRNN)    (None, 32)                2080      
                                                                 
Total params: 322,080
Trainable params: 322,080
Non-trainable params: 0
_________________________________________________________________


---

### Q3
```
tanh([-1, 0, 1]) = [a, b, c]
```

What can you say about ```a, b``` and ```c```?

**```tanh([-1, 0, 1])``` is of the form ```[-x, 0, x]``` where $\bf 0 < x < 1$. (In fact $\bf x = 0.76$)**

From the $\tanh$ graph, $\tanh(0) = 0$ and $\tanh(x) = -\tanh(-x)$. Also, $\tanh$ is contained between in the interval $[-1, 1]$.

---


### Q4
```
model = Sequential()
model.add(Embedding(input_dim=10000, output_dim=16))
model.add(LSTM(8))
```

```number_trainable_LSTM_parameters = ?```

**800**


An LSTM layer has four simple RNN's. 

`output = gated_tanh(memory)` and `memory = gated_prev_memory + gated_SimpleRNN`. There are three gates and a simple RNN – four RNNs in total.

Here N = 16 and M = 8. So Simple RNN parameters sum to `(16 + 8 + 1) * 8 = 200`. Then 4 * 200 = 80

In [18]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Embedding, LSTM
model = Sequential()
model.add(Embedding(10000, 16))
model.add(LSTM(8))
model.summary()

Model: "sequential_12"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_5 (Embedding)     (None, None, 16)          160000    
                                                                 
 lstm_2 (LSTM)               (None, 8)                 800       
                                                                 
Total params: 160,800
Trainable params: 160,800
Non-trainable params: 0
_________________________________________________________________


---